File: backend/app/schemas/payments.py
from datetime import date
from uuid import UUID

from pydantic import BaseModel


class PaymentRecordRequest(BaseModel):
    lease_id: UUID | None = None
    tenant_whatsapp: str | None = None
    amount: float
    payment_method: str | None = None
    reference: str | None = None
    paid_date: date | None = None


class PaymentResponse(BaseModel):
    id: UUID
    lease_id: UUID
    amount_due: float
    amount_paid: float | None
    due_date: date
    paid_date: date | None
    status: str
    payment_method: str | None
    notes: str | None
    tenant_name: str | None = None
    unit_identifier: str | None = None

    model_config = {"from_attributes": True}


class PaymentRecordResult(BaseModel):
    payment: PaymentResponse
    matched: bool
    receipt_sent: bool
    message: str


class CheckOverdueResult(BaseModel):
    processed: int
    reminders_sent: int
    landlord_alerts: int
    escalations: list[dict]


class LatePatternItem(BaseModel):
    tenant_name: str
    unit_identifier: str
    times_late: int
    avg_days_late: float


class PropertyBreakdownItem(BaseModel):
    unit_identifier: str
    expected: float
    collected: float
    status: str


class MonthlyReportResponse(BaseModel):
    landlord_id: UUID
    landlord_name: str
    period: str
    total_expected: float
    total_collected: float
    total_outstanding: float
    collection_rate: float
    payments_on_time: int
    payments_late: int
    payments_missed: int
    late_patterns: list[LatePatternItem]
    property_breakdown: list[PropertyBreakdownItem]
    active_payment_plans: int
    total_arrears_under_plan: float


class GenerateLeasePaymentsResult(BaseModel):
    created: int
    lease_id: str
    message: str

File: frontend/lib/api/payments.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export interface PaymentRecordRequest {
  lease_id?: string;
  tenant_whatsapp?: string;
  amount: number;
  payment_method?: string;
  reference?: string;
  paid_date?: string;
}

export interface PaymentRecordResult {
  payment: {
    id: string;
    lease_id: string;
    amount_due: number;
    amount_paid: number | null;
    due_date: string;
    paid_date: string | null;
    status: string;
    payment_method: string | null;
    notes: string | null;
    tenant_name: string | null;
    unit_identifier: string | null;
  };
  matched: boolean;
  receipt_sent: boolean;
  message: string;
}

export interface CheckOverdueResult {
  processed: number;
  reminders_sent: number;
  landlord_alerts: number;
  escalations: Array<{
    payment_id: string;
    tenant: string;
    unit: string;
    days_overdue: number;
    level: string;
    amount: number;
  }>;
}

export interface MonthlyReportResponse {
  landlord_id: string;
  landlord_name: string;
  period: string;
  total_expected: number;
  total_collected: number;
  total_outstanding: number;
  collection_rate: number;
  payments_on_time: number;
  payments_late: number;
  payments_missed: number;
  late_patterns: Array<{
    tenant_name: string;
    unit_identifier: string;
    times_late: number;
    avg_days_late: number;
  }>;
  property_breakdown: Array<{
    unit_identifier: string;
    expected: number;
    collected: number;
    status: string;
  }>;
  active_payment_plans: number;
  total_arrears_under_plan: number;
}

async function apiFetch<T>(path: string, options?: RequestInit): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...options,
  });
  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(error.detail || "API request failed");
  }
  return res.json();
}

export async function recordPayment(
  req: PaymentRecordRequest
): Promise<PaymentRecordResult> {
  return apiFetch("/payments/record", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export async function triggerOverdueCheck(): Promise<CheckOverdueResult> {
  return apiFetch("/payments/check-overdue", { method: "POST" });
}

export interface GenerateLeasePaymentsResult {
  created: number;
  lease_id: string;
  message: string;
}

export async function generatePaymentsForLease(leaseId: string): Promise<GenerateLeasePaymentsResult> {
  return apiFetch(`/payments/generate-for-lease/${leaseId}`, { method: "POST" });
}

export async function getMonthlyReport(
  landlordId: string,
  year?: number,
  month?: number
): Promise<MonthlyReportResponse> {
  const params = new URLSearchParams();
  if (year) params.set("year", String(year));
  if (month) params.set("month", String(month));
  const qs = params.toString();
  return apiFetch(`/payments/report/${landlordId}${qs ? "?" + qs : ""}`);
}
